File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_io_rw_test.ts
/**
 * g_hshk_read_io / g_hshk_write_io（IO→CPU 15h / 16h）
 * 根拠: HandShake.mdc「IO読み出し」「IO書き込み」
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const BASE_REGS = {
  R2: 0x2222,
  R3: 0x3333,
  R4: 0x4444,
} as const;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ handshake モックで 1 ケースを実行する。
 * @param fn 本体
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * IO が HSHK_IN_REQ を上げるまで待つ。
 * @param mock IO モック
 * @param timeoutMs 上限 ms
 */
async function waitReq1(
  mock: IoBoardHandshakeMock,
  timeoutMs = 2000,
): Promise<void> {
  const t0 = Date.now();
  while (mock.bus.HSHK_IN_REQ !== 1) {
    if (Date.now() - t0 > timeoutMs) {
      throw new Error("timeout waiting HSHK_IN_REQ");
    }
    await new Promise<void>((resolve) => setTimeout(resolve, 0));
  }
}

test("15h は指定バイト数＋status を返す", async () => {
  await withCase(async (s, mock) => {
    const toCpu = Uint8Array.from([0x15, 0x00, 0x00, 0xa0, 0x02, 0x00]);
    const io = mock.exchangeWithCpu(toCpu, 3);
    await waitReq1(mock);
    await s.call("g_handshake_interrupt_handler", {
      registers: { ...BASE_REGS },
    });
    const reply = await io;
    expect(reply.length).toBe(3);
    expect(reply[2]).toBe(0);
    s.expectRegisters({ R0: 0, R4: 0x4444 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_io_rw_test.rs
use std::sync::Arc;

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

fn base_regs() -> CallRegisters {
    CallRegisters {
        r2: Some(0x2222),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    s.run_init()?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

#[test]
fn cmd_15h_read_io_returns_requested_bytes_plus_status() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        let to_cpu = [0x15, 0x00, 0x00, 0xa0, 0x02, 0x00];
        let reply = mock.run_io_handler_exchange(&to_cpu, 3, || {
            let _ = s.call(
                "g_handshake_interrupt_handler",
                CallOptions {
                    registers: Some(base_regs()),
                    ..Default::default()
                },
            )?;
            Ok(())
        })?;
        assert_eq!(reply.len(), 3);
        assert_eq!(reply[2], 0);
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
